### PR TITLE
Added role module with rbac

### PR DESCRIPTION
**Pull request by bsdhrubo**
# Description

(Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.)

## Type of change

(Please delete options that are not relevant.)

- [ ] Link to a **GH Issue** or **Jira** task here for reference.
- [ ] Bug fix
- [x] New feature

# How Has This Been Tested?

(Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration)

- [x] Tested by myself
- [ ] Unit test

# Developer Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Review check list:

- [ ] Separation of Concerns followed (Reviewer name/NaN)
- [ ] Code is in sync with existing code patterns/technologies. (Reviewer name/NaN)
- [ ] Are there any obvious optimizations that will improve performance? (Reviewer name/NaN)
- [ ] Can any logging or debugging code be removed? (Reviewer name/NaN)
- [ ] Are all data inputs checked (for the correct type, length, format, and range) and encoded? (Reviewer name/NaN)
- [ ] Are invalid parameter values handled? (Reviewer name/NaN)
- [ ] Is error handling done the correct way? (Reviewer name/NaN)
- [ ] Are all the edge cases (null, 0, negative) handled properly? (Reviewer name/NaN)

### CODE REVIEW

File: packages/headless/src/database/mongodb/store-admin-role/store-admin.role.model.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    },
+    name: {
+      type: String,
+      trim: true,
+      required: true,
+    },
+    roleType:{
+      type: String,
+      index: true,
+      default: RoleTypeEnum.STORE_ADMIN
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/store-admin-role/store-admin.role.model.ts]: remove this field

File: packages/headless/src/database/mongodb/store-admin/store-admin.model.ts [hunk 2]
```diff
@@ -9,29 +9,67 @@ const StoreAdminSchema = new Schema<StoreAdmin>(
...
     },
-    // branchId
-    branch: {
+    storeId: {
+      type: String,
+      trim: true,
+      index: true,
+      required: true,
+    },
+    branchId: {
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/store-admin/store-admin.model.ts]: use plural name for multiple data storing

File: packages/headless/src/database/mongodb/store-admin/store-admin.model.ts [hunk 3]
```diff
@@ -9,29 +9,67 @@ const StoreAdminSchema = new Schema<StoreAdmin>(
...
+      required: true,
+    },
+    phone: {
       type: String,
       index: true,
+      default: '',
+    },
+    password: {
+      type: String,
+      index: true,
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/store-admin/store-admin.model.ts]: this should not go under indexing

File: packages/headless/src/database/mongodb/store/index.ts [hunk 4]
```diff
@@ -62,22 +66,60 @@ export class StoreDatabase implements IStoreDatabase {
...
+       * step 4: update admin with store id
+       * step 5: update role with store id
+       */
+      let role: Role = await new StoreAdminRoleModel({
+        ...data.role,
+        roleType: RoleTypeEnum.STORE_ADMIN,
+        storeId: 'temp',
+      }).save({ session });
+
+      let storeAdmin: StoreAdmin = await new StoreAdminModel({
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/store/index.ts]: change the flow. replace "temp" by randomuuid.

File: packages/headless/src/database/mongodb/super-admin-role/super-admin.role.model.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import { randomUUID } from 'crypto';
+import { RoleTypeEnum } from 'models';
+
+const SuperAdminRoleSchema = new Schema<Role>(
+  {
+    id: {
+      type: String,
+      default: () => randomUUID(),
+      unique: true,
+      indexe: true,
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/super-admin-role/super-admin.role.model.ts]: spelling mistake. and unique means unique index. no need to rewrite index

File: packages/headless/src/database/mongodb/super-admin-role/super-admin.role.model.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      type: String,
+      trim: true,
+      unique: true,
+      index: true,
+      required: true,
+    },
+    roleType: {
+      type: String,
+      index: true,
+      default: RoleTypeEnum.SUPER_ADMIN
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/super-admin-role/super-admin.role.model.ts]: remove this type

File: packages/headless/src/database/mongodb/super-admin/super-admin.model.ts [hunk 7]
```diff
@@ -12,51 +12,74 @@ const SuperAdminSchema = new Schema<SuperAdmin>(
...
     phone: {
       type: String,
       index: true,
-      default: ''
+      default: '',
     },
     password: {
       type: String,
-      required: true
+      index: true,
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/super-admin/super-admin.model.ts]: remove this index

File: packages/headless/src/guards/admin-role.guard.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    if (!requiredPermissions?.length) {
+      return true;
+    }
+
+    const user = context.switchToHttp().getRequest()?.user;
+
+    // allow access to super admin
+    if (user.role === 'super-admin') {
+      return true;
+    }
```
Comment (Anjarul-Islam) [inline packages/headless/src/guards/admin-role.guard.ts:28]: Add a TODO: comment if can't complete this now

File: packages/headless/src/database/mongodb/store-admin/store-admin.model.ts (hunk 2 above)
Comment (bsdhrubo) [reply packages/headless/src/database/mongodb/store-admin/store-admin.model.ts]: Done

File: packages/headless/src/database/mongodb/store-admin/store-admin.model.ts (hunk 3 above)
Comment (bsdhrubo) [reply packages/headless/src/database/mongodb/store-admin/store-admin.model.ts]: Done

File: packages/headless/src/database/mongodb/store/index.ts (hunk 4 above)
Comment (bsdhrubo) [reply packages/headless/src/database/mongodb/store/index.ts]: Done

File: packages/headless/src/database/mongodb/super-admin-role/super-admin.role.model.ts (hunk 5 above)
Comment (bsdhrubo) [reply packages/headless/src/database/mongodb/super-admin-role/super-admin.role.model.ts]: Fixed

File: packages/headless/src/database/mongodb/super-admin/super-admin.model.ts (hunk 7 above)
Comment (bsdhrubo) [reply packages/headless/src/database/mongodb/super-admin/super-admin.model.ts]: Done

File: packages/headless/src/guards/admin-role.guard.ts (hunk 8 above)
Comment (bsdhrubo) [reply packages/headless/src/guards/admin-role.guard.ts:28]: Done

File: packages/headless/src/modules/store-admin-role/services/index.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    body: UpdateRoleDto,
+  ): Promise<IServiceResponse<Role>> {
+    const isExist = await this.roleRepository.findOne({
+      id: body.id,
+      storeId: adminInfo.storeId,
+    });
+    if (isExist) {
+      if (isExist.name === body?.name) {
+        return errorResponse(
+          'This role is already exist!',
```
Comment (Anjarul-Islam) [inline packages/headless/src/modules/store-admin-role/services/index.ts]: nested branching is not good coding style. this creates clunky code. rather you could do something like
 ```
if (!isExist) {
        return ...
 }
other logic....
```

Comment (bsdhrubo) [reply packages/headless/src/modules/store-admin-role/services/index.ts]: Fixed

**Review by Anjarul-Islam — APPROVED** [review]

**Pull request merged**